messaging/src/main/java/org/axonframework/configuration/NewConfigurer.java

```diff
@@ -0,0 +1,209 @@
+ * {@link #registerDecorator(Class, int, ComponentDecorator) decorators} of these components, check if a component
+ * {@link #hasComponent(Class) exists}, register {@link #registerEnhancer(ConfigurerEnhancer) enhancers} for the entire
+ * configurer, and {@link #registerModule(ModuleBuilder) modules}.
+ *
+ * @param <S> The type of configurer this implementation returns. This generic allows us to support fluent interfacing.
+ * @author Allard Buijze
+ * @author Steven van Beelen
+ * @since 3.0.0
+ */
+// TODO Rename to Configurer once the old Configurer is removed
+public interface NewConfigurer<S extends NewConfigurer<S>> extends LifecycleOperations {
+
+    /**
+     * Registers a {@link Component} that should be made available to other {@link Component components} or
+     * {@link Module modules} in the {@link NewConfiguration} that this {@code Configurer} will result in.
+     * <p>
+     * The given {@code builder} function gets the {@link NewConfiguration configuration} as input, and is expected to
+     * provide the component as output. The component will be registered under an {@link Identifier} based on the given
+     * {@code type}.
+     * <p>
+     * Note that registering a component twice for the same {@code type} will remove the previous registration!
+     *
+     * @param type    The declared type of the component to build, typically an interface.
+     * @param builder The builder function of this component.
+     * @param <C>     The type of component the {@code builder} builds.
+     * @return The current instance of the {@code Configurer} for a fluent API.
+     */
+    default <C> S registerComponent(@Nonnull Class<C> type,
+                                    @Nonnull ComponentBuilder<C> builder) {
+        return registerComponent(type, type.getSimpleName(), builder);
+    }
+
+    /**
+     * Registers a {@link Component} that should be made available to other {@link Component components} or
+     * {@link Module modules} in the {@link NewConfiguration} that this {@code Configurer} will result in.
+     * <p>
+     * The given {@code builder} function gets the {@link NewConfiguration configuration} as input, and is expected to
+     * provide the component as output. The component will be registered under an {@link Identifier} based on the given
+     * {@code type} and {@code name} combination.
+     * <p>
+     * Note that registering a component twice for the same {@code type} and {@code name} will remove the previous
+     * registration!
+     *
+     * @param type    The declared type of the component to build, typically an interface.
+     * @param name    The name of the component to build.
+     * @param builder The builder function of this component.
+     * @param <C>     The type of component the {@code builder} builds.
+     * @return The current instance of the {@code Configurer} for a fluent API.
+     */
+    <C> S registerComponent(@Nonnull Class<C> type,
```

Comment:
Question for reviewers: Should we provide a register method that explicitely states to override the `ComponetBuilder`? Right now, it just does this, but we could have a `boolean override`, defaulting to X. The `Component` class is already able to support this through the `Component#update` operation. Wdyt?

Reply:
Ah interesting, one of my review comments is about this `update ` operation.

We need to decide whether it is normal for a user to call register for a component that already exists. 
As the `MessagingConfigurer` will provide defaults that can be overridden, I think the chances are very big a user will do it. Because of the defaults, any call to this method is an override, and will result in users always passing in a `true` option, making it redundant.
So I think we should not put in a boolean

Reply:
Alright, thanks for your 2-cents! 

I have (of course) also had a chat about this with @abuijze, which referred to Spring's recent change to fail if there are several beans with the same name. 
If we map that idea to our environment, we would have several components with the same name "silently" overriding one another.
Although not really silent (the `AbstractConfigurer` logs it on WARN-level), perhaps it is "nicer" to simply fail?
Or, thus, to provide an override value, that we either enable or disable by default for a register component operation.

Does this bit of information perhaps change your opinion, @CodeDrivenMitch?

Reply:
Yes, let's discuss this in the AF5 sessions in further detail.